include/ghex/unstructured/pattern.hpp

```diff
@@ -0,0 +1,346 @@
+/*
+ * GridTools
+ *
+ * Copyright (c) 2014-2019, ETH Zurich
+ * All rights reserved.
+ *
+ * Please, refer to the LICENSE file in the root directory.
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+#ifndef INCLUDED_GHEX_UNSTRUCTURED_PATTERN_HPP
+#define INCLUDED_GHEX_UNSTRUCTURED_PATTERN_HPP
+
+#include <vector>
+#include <map>
+#include <cassert>
+#include <numeric>
+#include <cstring>
+
+#include "../transport_layer/communicator.hpp"
+#include "../transport_layer/mpi/setup.hpp"
+#include "../pattern.hpp"
+#include "../buffer_info.hpp"
+#include "./grid.hpp"
+#include "../allocator/unified_memory_allocator.hpp"
+
+namespace gridtools {
+    namespace ghex {
+
+        /** @brief unstructured pattern specialization
+         *
+         * This class provides access to the receive and send iteration spaces, determined by the halos,
+         * and holds all connections to the neighbors.
+         *
+         * @tparam Transport transport protocol
+         * @tparam Index index type for domain and iteration space
+         * @tparam DomainId domain id type*/
+        template<typename Transport, typename Index, typename DomainId>
+        class pattern<Transport, unstructured::detail::grid<Index>, DomainId> {
+
+            public:
+
+                // member types
+
+                using communicator_type = tl::communicator<Transport>;
+                using address_type = typename communicator_type::address_type;
+                using index_type = Index;
+                using grid_type = unstructured::detail::grid<index_type>;
+                using domain_id_type = DomainId;
+                using pattern_container_type = pattern_container<Transport, grid_type, domain_id_type>;
+
+                friend class pattern_container<Transport, grid_type, domain_id_type>;
+
+                /** @brief essentially a partition index and a sequence of remote indexes;
+                 * number of levels is provided as well, defaul is 1 (2D case):
+                 * the assumption is that each 2D element is a column, with 'levels' vertical elements.
+                 * WARN: a second index, and therefore a more complex iteration space,
+                 * is needed to handle correctly multiple vertical layers*/
+                class iteration_space {
+
+                    private:
+
+                        int m_partition;
+                        std::vector<index_type, gridtools::ghex::allocator::unified_memory_allocator<index_type>> m_local_index;
+                        std::size_t m_levels;
+
+                    public:
+
+                        // ctors
+                        iteration_space() noexcept = default;
+                        iteration_space(const int partition,
+                                const std::vector<index_type, gridtools::ghex::allocator::unified_memory_allocator<index_type>>& local_index,
+                                const std::size_t levels = 1) noexcept :
```

Comment:
I think this can throw

Reply:
Why only this one and not the following?

Reply:
I think the move constructor of the vector does not throw. If it does, then also the next one can throw

Reply:
Ok, done.